### DISCUSSION TITLE
Best way to pause ongoing read/writes to allow other uses of the SD card?

### DISCUSSION THREAD

**Discussion by saybur** [ZuluSCSI/ZuluSCSI-firmware#193] [Q&A] [answered]
I've been doing some experimentation with audio output on a ZuluSCSI RP2040 and have managed to get a functional prototype that can play samples off the memory card during light SCSI usage via a (mis)use of recurring platform_reset_watchdog() calls. I'd like to develop this into something more robust, but have been running into issues wrapping my head around the best way to temporarily pause SCSI data transfers so audio sample streaming doesn't stall.

Does anyone have any thoughts about the best way this could be accomplished? I'm not averse to doing some experimentation, but I was hoping someone with a better grasp of the codebase might have some pointers about the best/safest place this could be done to avoid causing issues with SCSI transactions.

**Comment by PetteriAimonen** [comment, marked as answer]
On the RP2040 platform, there is actually mbed RTOS running. That could potentially help a bit, though currently SD card can only be accessed from the main thread.

`scsiIsWriteFinished()` and `scsiIsReadFinished()` are the polling calls for SCSI read & write. They could be hooked in platform code, though that will need a separate status in `sd_card_sdio.cpp` to make sure we are not currently in SD data callback. Or we could add a separate poll function for handling audio, which would be called from `ZuluSCSI_disk.cpp` only when SD card is free.

Longest SD card reads and writes will be 64 kB, resulting in normal poll interval of around 5 ms.

The longest pauses will be during continuous SD card writing. The SD spec says that cards can pause for up to 100 ms to do internal erase operations, but I have seen cards pause up to 500 ms in practice. So there should be enough audio data buffered for that, which is around 80kB. There is not quite that much RAM available currently, but this is because `rp2040.ld` puts most code in RAM. Even with all code in flash, performance is almost as good, and some more careful selection of critical functions should free up RAM.

On the other hand, if the audio stutters a bit with slowest SD cards when doing continuous writing, that is not very critical. For a CD-only emulation writing wouldn't be an issue, but the same device can be emulating a disk drive at the same time.

**Reply by PetteriAimonen** [reply]
I pushed branch `dev_platform_poll` that adds `platform_poll()` function that may be helpful:
https://github.com/ZuluSCSI/ZuluSCSI-firmware/tree/dev_platform_poll

On further consideration, another critical point is responding to SCSI selections on platforms that have very short selection timeout. I think at least MacPlus boot ROM does this, timing out in 1 ms while spec recommends 250 ms timeout. It may be that we need to move the BSY response to selection interrupt in RP2040 platform - shouldn't be too difficult.

**Comment by saybur** [comment]
Wow, this is perfect! I'll implement using the new platform call, that is exactly what I needed. Thank you!

I think you're right about the stuttering, a bit here and there during heavy operations and/or with slower cards is tolerable. I'll test once I have a proper implementation and see how frequent it ends up being. More testing/tweaking should also give a better idea of how much overall buffering can be done with the existing memory available: ideally I'd like to propose a minimal PR and wait for memory reduction down the line if audio playback proves workable. Some of this may wait for #187 as well, given the audio data connection.

As for compatibility with the short selection window, I'm not aware of CD-based software that would want to play audio on a Plus, though it certainly might be out there. If this does end up implemented maybe it could be handled with a log warning (or auto-disabling) when "MacPlus" is set as the system type.
